[flaky test] query with pipeline breaker seems unstable

see: https://github.com/apache/pinot/actions/runs/5147527036/jobs/9268030409

```
2023-06-01T18:30:47.9165106Z [ERROR] org.apache.pinot.query.runtime.queries.ResourceBasedQueriesTest.testQueryTestCasesWithH2[hint_option_queries, SELECT /*+ joinOptions(join_strategy='dynamic_broadcast', is_colocated_by_join_keys='true') */ hint_option_queries_tbl1.num, hint_option_queries_tbl1.name FROM hint_option_queries_tbl1 WHERE hint_option_queries_tbl1.num IN (SELECT hint_option_queries_tbl2.num FROM hint_option_queries_tbl2 WHERE hint_option_queries_tbl2.val IN ('xxx', 'yyy')), SELECT /*+ joinOptions(join_strategy='dynamic_broadcast', is_colocated_by_join_keys='true') */ hint_option_queries_tbl1.num, hint_option_queries_tbl1.name FROM hint_option_queries_tbl1 WHERE hint_option_queries_tbl1.num IN (SELECT hint_option_queries_tbl2.num FROM hint_option_queries_tbl2 WHERE hint_option_queries_tbl2.val IN ('xxx', 'yyy')), null, false](226)  Time elapsed: 0.216 s  <<< FAILURE!
2023-06-01T18:30:47.9166403Z java.lang.RuntimeException: 
2023-06-01T18:30:47.9166782Z Received error query execution result block: {1000=Cancelled by sender with exception: Query was cancelled!
2023-06-01T18:30:47.9167262Z java.lang.RuntimeException: Cancelled by sender with exception: Query was cancelled!
2023-06-01T18:30:47.9167741Z 	at org.apache.pinot.query.mailbox.GrpcSendingMailbox.cancel(GrpcSendingMailbox.java:84)
2023-06-01T18:30:47.9168305Z 	at org.apache.pinot.query.runtime.operator.exchange.BlockExchange.cancel(BlockExchange.java:173)
2023-06-01T18:30:47.9168904Z 	at org.apache.pinot.query.runtime.operator.MailboxSendOperator.cancel(MailboxSendOperator.java:182)
2023-06-01T18:30:47.9169494Z 	at org.apache.pinot.query.runtime.operator.OpChain.cancel(OpChain.java:79)
2023-06-01T18:30:47.9169820Z ...
2023-06-01T18:30:47.9170098Z Caused by: java.lang.InterruptedException: Query was cancelled!
2023-06-01T18:30:47.9170598Z 	at org.apache.pinot.query.runtime.executor.OpChainSchedulerService$1.runJob(OpChainSchedulerService.java:88)
2023-06-01T18:30:47.9171039Z 	... 6 more}
2023-06-01T18:30:47.9171479Z 	at org.apache.pinot.query.service.dispatch.QueryDispatcher.reduceMailboxReceive(QueryDispatcher.java:223)
2023-06-01T18:30:47.9172105Z 	at org.apache.pinot.query.service.dispatch.QueryDispatcher.runReducer(QueryDispatcher.java:204)
2023-06-01T18:30:47.9172681Z 	at org.apache.pinot.query.runtime.QueryRunnerTestBase.queryRunner(QueryRunnerTestBase.java:125)
2023-06-01T18:30:47.9173300Z 	at org.apache.pinot.query.runtime.queries.ResourceBasedQueriesTest.runQuery(ResourceBasedQueriesTest.java:298)
2023-06-01T18:30:47.9174026Z 	at org.apache.pinot.query.runtime.queries.ResourceBasedQueriesTest.testQueryTestCasesWithH2(ResourceBasedQueriesTest.java:223)
2023-06-01T18:30:47.9174641Z 	at jdk.internal.reflect.GeneratedMethodAccessor31.invoke(Unknown Source)
2023-06-01T18:30:47.9175181Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-06-01T18:30:47.9175667Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2023-06-01T18:30:47.9176121Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:108)
2023-06-01T18:30:47.9176582Z 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:661)
2023-06-01T18:30:47.9193016Z 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:869)
2023-06-01T18:30:47.9193519Z 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1193)
2023-06-01T18:30:47.9193997Z 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:126)
2023-06-01T18:30:47.9194470Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
2023-06-01T18:30:47.9194867Z 	at org.testng.TestRunner.privateRun(TestRunner.java:744)
2023-06-01T18:30:47.9195208Z 	at org.testng.TestRunner.run(TestRunner.java:602)
2023-06-01T18:30:47.9195547Z 	at org.testng.SuiteRunner.runTest(SuiteRunner.java:380)
2023-06-01T18:30:47.9195922Z 	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:375)
2023-06-01T18:30:47.9196307Z 	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
2023-06-01T18:30:47.9196658Z 	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
2023-06-01T18:30:47.9197033Z 	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
2023-06-01T18:30:47.9197429Z 	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
2023-06-01T18:30:47.9197815Z 	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1301)
2023-06-01T18:30:47.9198347Z 	at org.testng.TestNG.runSuitesLocally(TestNG.java:1226)
2023-06-01T18:30:47.9198683Z 	at org.testng.TestNG.runSuites(TestNG.java:1144)
2023-06-01T18:30:47.9198992Z 	at org.testng.TestNG.run(TestNG.java:1115)
2023-06-01T18:30:47.9199392Z 	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:136)
2023-06-01T18:30:47.9199990Z 	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:112)
2023-06-01T18:30:47.9200654Z 	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:99)
2023-06-01T18:30:47.9201205Z 	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:145)
2023-06-01T18:30:47.9201808Z 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
2023-06-01T18:30:47.9202332Z 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
2023-06-01T18:30:47.9202797Z 	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
2023-06-01T18:30:47.9203253Z 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
```

seems like there are some distributed deadlock? will investigate if i can find another example.

CC @Jackie-Jiang

--- a/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/RoundRobinScheduler.java
+++ b/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/RoundRobinScheduler.java
@@ -169,6 +169,8 @@ public class RoundRobinScheduler implements OpChainScheduler {
     } finally {
       _lock.unlock();
     }
+    // invoke opChain deregister callback
+    operatorChain.getOpChainFinishCallback().accept(operatorChain.getId());
   }
 
   @Override

--- a/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OpChain.java
+++ b/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OpChain.java
@@ -19,6 +19,7 @@
 package org.apache.pinot.query.runtime.operator;
 
 import java.util.List;
+import java.util.function.Consumer;
 import org.apache.pinot.core.common.Operator;
 import org.apache.pinot.query.runtime.blocks.TransferableBlock;
 import org.apache.pinot.query.runtime.plan.OpChainExecutionContext;
@@ -33,12 +34,19 @@ public class OpChain implements AutoCloseable {
   private final List<String> _receivingMailboxIds;
   private final OpChainId _id;
   private final OpChainStats _stats;
+  private final Consumer<OpChainId> _opChainFinishCallback;
 
   public OpChain(OpChainExecutionContext context, MultiStageOperator root, List<String> receivingMailboxIds) {
+    this(context, root, receivingMailboxIds, (id) -> { });
+  }
+
+  public OpChain(OpChainExecutionContext context, MultiStageOperator root, List<String> receivingMailboxIds,
+      Consumer<OpChainId> opChainFinishCallback) {
     _root = root;
     _receivingMailboxIds = receivingMailboxIds;
     _id = context.getId();
     _stats = context.getStats();
+    _opChainFinishCallback = opChainFinishCallback;
   }
 
   public Operator<TransferableBlock> getRoot() {
@@ -47,6 +55,10 @@ public class OpChain implements AutoCloseable {
 
   public List<String> getReceivingMailboxIds() {
     return _receivingMailboxIds;
+  }
+
+  public Consumer<OpChainId> getOpChainFinishCallback() {
+    return _opChainFinishCallback;
   }
 
   public OpChainId getId() {

--- a/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/pipeline/PipelineBreakerExecutor.java
+++ b/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/pipeline/PipelineBreakerExecutor.java
@@ -18,9 +18,12 @@
  */
 package org.apache.pinot.query.runtime.plan.pipeline;
 
+import java.io.IOException;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
 import org.apache.pinot.core.common.Operator;
 import org.apache.pinot.query.mailbox.MailboxService;
 import org.apache.pinot.query.planner.plannode.MailboxReceiveNode;
@@ -31,12 +34,15 @@ import org.apache.pinot.query.runtime.operator.OpChain;
 import org.apache.pinot.query.runtime.plan.DistributedStagePlan;
 import org.apache.pinot.query.runtime.plan.PhysicalPlanContext;
 import org.apache.pinot.query.runtime.plan.PhysicalPlanVisitor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 /**
  * Utility class to run pipeline breaker execution and collects the results.
  */
 public class PipelineBreakerExecutor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PipelineBreakerExecutor.class);
   private PipelineBreakerExecutor() {
     // do not instantiate.
   }
@@ -92,10 +98,15 @@ public class PipelineBreakerExecutor {
       throws Exception {
     PipelineBreakerOperator pipelineBreakerOperator = new PipelineBreakerOperator(
         physicalPlanContext.getOpChainExecutionContext(), pipelineWorkerMap);
+    CountDownLatch latch = new CountDownLatch(1);
     OpChain pipelineBreakerOpChain = new OpChain(physicalPlanContext.getOpChainExecutionContext(),
-        pipelineBreakerOperator,
-        physicalPlanContext.getReceivingMailboxIds());
+        pipelineBreakerOperator, physicalPlanContext.getReceivingMailboxIds(), (id) -> latch.countDown());
     scheduler.register(pipelineBreakerOpChain);
-    return pipelineBreakerOperator.getResult();
+    long timeoutMs = physicalPlanContext.getDeadlineMs() - System.currentTimeMillis();
+    if (latch.await(timeoutMs, TimeUnit.MILLISECONDS)) {
+      return pipelineBreakerOperator.getResultMap();
+    } else {
+      throw new IOException("Exception occur when awaiting breaker results!");
+    }
   }
 }

--- a/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/pipeline/PipelineBreakerOperator.java
+++ b/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/pipeline/PipelineBreakerOperator.java
@@ -18,14 +18,14 @@
  */
 package org.apache.pinot.query.runtime.plan.pipeline;
 
+import com.google.common.collect.ImmutableSet;
 import java.util.ArrayDeque;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.Deque;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.TimeUnit;
 import javax.annotation.Nullable;
 import org.apache.pinot.common.exception.QueryException;
 import org.apache.pinot.core.common.Operator;
@@ -39,7 +39,7 @@ public class PipelineBreakerOperator extends MultiStageOperator {
   private static final String EXPLAIN_NAME = "PIPELINE_BREAKER";
   private final Deque<Map.Entry<Integer, Operator<TransferableBlock>>> _workerEntries;
   private final Map<Integer, List<TransferableBlock>> _resultMap;
-  private final CountDownLatch _workerDoneLatch;
+  private final ImmutableSet<Integer> _expectedKeySet;
   private TransferableBlock _errorBlock;
 
 
@@ -47,20 +47,13 @@ public class PipelineBreakerOperator extends MultiStageOperator {
       Map<Integer, Operator<TransferableBlock>> pipelineWorkerMap) {
     super(context);
     _resultMap = new HashMap<>();
+    _expectedKeySet = ImmutableSet.copyOf(pipelineWorkerMap.keySet());
     _workerEntries = new ArrayDeque<>();
     _workerEntries.addAll(pipelineWorkerMap.entrySet());
-    _workerDoneLatch = new CountDownLatch(1);
   }
 
-  public Map<Integer, List<TransferableBlock>> getResult()
-      throws Exception {
-    boolean isWorkerDone =
-        _workerDoneLatch.await(_context.getDeadlineMs() - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
-    if (isWorkerDone && _errorBlock == null) {
-      return _resultMap;
-    } else {
-      throw new RuntimeException("Unable to construct pipeline breaker results due to timeout.");
-    }
+  public Map<Integer, List<TransferableBlock>> getResultMap() {
+    return _resultMap;
   }
 
   @Nullable
@@ -73,7 +66,7 @@ public class PipelineBreakerOperator extends MultiStageOperator {
   protected TransferableBlock getNextBlock() {
     if (System.currentTimeMillis() > _context.getDeadlineMs()) {
       _errorBlock = TransferableBlockUtils.getErrorTransferableBlock(QueryException.EXECUTION_TIMEOUT_ERROR);
-      _workerDoneLatch.countDown();
+      constructErrorResponse(_errorBlock);
       return _errorBlock;
     }
 
@@ -96,7 +89,7 @@ public class PipelineBreakerOperator extends MultiStageOperator {
       if (block != null) {
         if (block.isErrorBlock()) {
           _errorBlock = block;
-          _workerDoneLatch.countDown();
+          constructErrorResponse(block);
           return _errorBlock;
         }
         List<TransferableBlock> blockList = _resultMap.computeIfAbsent(worker.getKey(), (k) -> new ArrayList<>());
@@ -108,11 +101,18 @@ public class PipelineBreakerOperator extends MultiStageOperator {
     }
 
     if (_workerEntries.isEmpty()) {
-      // NOTIFY results are ready.
-      _workerDoneLatch.countDown();
       return TransferableBlockUtils.getEndOfStreamTransferableBlock();
     } else {
       return TransferableBlockUtils.getNoOpTransferableBlock();
+    }
+  }
+
+  /**
+   * Setting all result map to error if any of the pipeline breaker returns an ERROR.
+   */
+  private void constructErrorResponse(TransferableBlock errorBlock) {
+    for (int key : _expectedKeySet) {
+      _resultMap.put(key, Collections.singletonList(errorBlock));
     }
   }
 }

